flexdll.c

```diff
@@ -47,8 +47,99 @@ typedef struct dlunit {
 } dlunit;
 typedef void *resolver(void*, const char*);
 
-static int error = 0;
-static char error_buffer[256];
+/* Error reporting */
+/* The latest error must be kept in some variable so that flexdll_dlerror can
+ * report it but this causes data races (and possible segmentation faults) in
+ * multithreaded programs if that variable is global. So this must use
+ * thread-local storage instead.
+ *
+ * To ensure compatibility, the implementation does not require compiler support
+ * for thread-local storage (__thread) and instead relies on functions
+ * TlsGetValue, TlsSetValue, etc.
+ * This makes for an implementation a bit complicated, as the goal is to deal
+ * with errors and those functions do modify the result of further calls to
+ * GetLastError even when they succeed.
+ *
+ * This implementation is structured around the function get_tls_error that
+ * returns a structure containing, for the current thread, the code (and a
+ * buffer for the corresponding message) of the last flexdll error and the last
+ * system error, as returned by GetLastError, when relevant. It accepts an
+ * argument to decide what should be the behaviour regarding the current latest
+ * system error, depending on the entrypoint:
+ *
+ * - TLS_ERROR_RESET_LAST will reset what is stored in the structure, so this is
+ *   intended for initialisation entry points (flexdll_dlopen, flexdll_relocate)
+ * - TLS_ERROR_SAVE_LAST will record in last_error the current latest system
+ *   error before it is overriden by TlsGetValue (flexdll_dlerror)
+ * - TLS_ERROR_KEEP_LAST will record in last_error the current latest system
+ *   error unless last_error already contains an error (ll_dlerror, since this
+ *   can be called both from flexdll_dlerror (last_error already set) or exactly
+ *   when the error is encountered, in relocate)
+ *
+ * The other exported entrypoints do not need to access the error storage.
+ */
+
+typedef struct error_s {
+  int code;
+  int last_error;
+  char message[256];
+} err_t;
+
+#define TLS_ERROR_RESET_LAST 1
+#define TLS_ERROR_KEEP_LAST 2
+#define TLS_ERROR_SAVE_LAST 3
+
+static err_t *get_tls_error(int op) {
+  static volatile DWORD error_idx = TLS_OUT_OF_INDEXES;
+  DWORD new_idx, last_error;
+  err_t *error;
+
+  if(op == TLS_ERROR_SAVE_LAST || op == TLS_ERROR_KEEP_LAST)
+    last_error = GetLastError();
+
+retry:
+  if(error_idx == TLS_OUT_OF_INDEXES) {
+    new_idx = TlsAlloc();
+    if(new_idx == TLS_OUT_OF_INDEXES)
+      /* If we cannot allocate the way to report errors... */
+      return NULL;
+    /* According to documentation DWORD and LONG take both 32 bits so
+     * this uses InterlockedCompareExchange to store a DWORD */
+    if (InterlockedCompareExchange((LONG*)&error_idx, (LONG)new_idx, (LONG)TLS_OUT_OF_INDEXES) != (LONG)TLS_OUT_OF_INDEXES) {
+      if(!TlsFree(new_idx))
+        return NULL;
+      goto retry;
```

Comment:
If we reach this `goto`, can it happen that the condition at the toop of the loop `error_idx == TLS_OUT_OF_INDEXES` holds? If not, can we get rid of the `goto` altogether?

Reply:
Good point!
I didn't find anything in the documentation that would invite to retest and local tests seem to work. I had followed this [example](https://github.com/dra27/flexdll/commit/7bbaf6649922d011df7efc24f9c9427bb4bc19c9?diff=unified#diff-6da5df65a6271e4d53ced90f2bb6d89edcdf56ddf1c1c13d79b007ed572fbfdcR400) but that commit was explicitly and intentionally over-defensive, so I also think we can simplify.